### PR TITLE
feat: remove connection to users after backends stop federating WPB-2220

### PR DESCRIPTION
**Pull request by MarcinR**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2220" title="WPB-2220" target="_blank">TaskWPB-2220</a>  [iOS] Handle connection requests to users on backends with which we don’t federate
  </td></table>
  <br />
 

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability

----

# What's new in this PR?
There might be a situation when user can open profile of user whose backend has stopped federating with our backend.  If device owner tries to connect with that user backend will respond with an error. In this PR I'm handling this error and display message to the user

### CODE REVIEW

**Comment by KaterinaWire** [comment]
@MarcinR Please add a description to this PR? It's hard to review a PR if you're not in context.

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileActionsFactory.swift [hunk 1]
```diff
@@ -190,7 +191,7 @@ final class ProfileActionsFactory {
                 actions.append(.cancelConnectionRequest)
             } else if (user.isConnected && !user.hasEmptyName) || isOnSameTeam {
                 actions.append(.openOneToOne)
-            } else if user.canBeConnected && !user.isPendingApprovalBySelfUser {
+            } else if user.canBeConnected && !user.isPendingApprovalBySelfUser && isUserFormClassifiedBackend(user: user) {
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileActionsFactory.swift]: Why do we need this check?

File: wire-ios/Wire-iOS/Sources/Components/PopUpViewController.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+
+class PopUpViewController: UIViewController {
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/Components/PopUpViewController.swift]: We have `UIAlertController.showErrorAlert(title: String, message: String) ` alert with Ok button. Let's reuse existing code. And please test it on iPad.

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileViewController.swift [hunk 3]
```diff
@@ -547,6 +547,15 @@ extension ProfileViewController: ProfileViewControllerViewModelDelegate {
...
     func presentError(_ error: LocalizedError) {
+        if let connectionError = error as? ConnectToUserError,
+           connectionError == .federationDenied {
+            typealias Strings = L10n.Localizable.Error.Connection
+            let message = Strings.federationDeniedMessage(viewModel.user.name ?? "")
+            let title = Strings.federationDeniedTitle
+            let popupViewController = PopUpViewController(title: title, message: message)
+            present(popupViewController, animated: true)
+            return
+        }
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileViewController.swift]: suggestion:
func presentError(_ error: LocalizedError) {
typealias ErrorConnection = L10n.Localizable.Error.Connection

if let connectionError = error as? ConnectToUserError,
           connectionError == .federationDenied {
presentAlertWithOKButton(
            title: ErrorConnection.title,
            message: ErrorConnection.message
        )
} else {
presentLocalizedErrorAlert(
}

File: wire-ios/Wire-iOS/Sources/Components/PopUpViewController.swift (hunk 2 above)
Comment (MarcinR) [reply wire-ios/Wire-iOS/Sources/Components/PopUpViewController.swift]: but that doesn't match whats on figma

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileActionsFactory.swift (hunk 1 above)
Comment (MarcinR) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileActionsFactory.swift]: to hide "connect" button if backends not federating

File: wire-ios-request-strategy/Sources/Request Strategies/Connection/Actions/UpdateConnectionActionHandlerTests.swift [hunk 4]
```diff
@@ -80,7 +80,8 @@ class UpdateConnectionActionHandlerTests: MessagingTestBase {
             (.connectionLimitReached, responseFailure(code: 403, label: .connectionLimit, apiVersion: .v0)),
             (.noIdentity, responseFailure(code: 403, label: .noIdentity, apiVersion: .v0)),
             (.missingLegalholdConsent, responseFailure(code: 403, label: .missingLegalholdConsent, apiVersion: .v0)),
-            (.notConnected, responseFailure(code: 403, label: .notConnected, apiVersion: .v0))
+            (.notConnected, responseFailure(code: 403, label: .notConnected, apiVersion: .v0)),
+            (.federationDenied, responseFailure(code: 400, label: .federationDenied, apiVersion: .v0)),
```
Comment (johnxnguyen) [inline wire-ios-request-strategy/Sources/Request Strategies/Connection/Actions/UpdateConnectionActionHandlerTests.swift]: Should it be v4?

**Review by David-Henner — APPROVED** [review]

File: wire-ios-data-model/Source/Model/Connection/ZMConnection+Actions.swift [hunk 5]
```diff
@@ -24,6 +24,7 @@ public enum ConnectToUserError: Error {
     case connectionLimitReached
     case missingLegalholdConsent
     case internalInconsistency
+    case invalidDomain
```
Comment (KaterinaWire) [inline wire-ios-data-model/Source/Model/Connection/ZMConnection+Actions.swift]: The tech spec was changed, we don't use this label. Have you tried to run the app and check what was deployed?

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileActionsFactory.swift [hunk 6]
```diff
@@ -19,6 +19,7 @@
 import Foundation
 import WireCommonComponents
 import WireDataModel
+import WireSyncEngine
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileActionsFactory.swift]: I don't think we need this?

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileViewController.swift [hunk 7]
```diff
@@ -547,6 +547,14 @@ extension ProfileViewController: ProfileViewControllerViewModelDelegate {
     }
 
     func presentError(_ error: LocalizedError) {
-        presentLocalizedErrorAlert(error)
+        typealias Strings = L10n.Localizable.Error.Connection
+
+        if let connectionError = error as? ConnectToUserError,
+           connectionError == .invalidDomain {
+            let title = Strings.federationDeniedMessage(viewModel.user.name ?? "")
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileViewController.swift]: Shouldn't this string be a message in the alert?

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileViewController.swift [hunk 8]
```diff
@@ -547,6 +547,14 @@ extension ProfileViewController: ProfileViewControllerViewModelDelegate {
     }
 
     func presentError(_ error: LocalizedError) {
-        presentLocalizedErrorAlert(error)
+        typealias Strings = L10n.Localizable.Error.Connection
+
+        if let connectionError = error as? ConnectToUserError,
+           connectionError == .invalidDomain {
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/ProfileViewController.swift]: Please double check, should be another label.

**Review by KaterinaWire — APPROVED** [review]

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**